### PR TITLE
Fixes #60939, make the attention capable to handle cases in fp16

### LINKED ISSUE #60939: Functional multi_head_attention_forward softmax get nan for fp16 mode
## 🐛 Bug

In torch.nn.functional, the function multi_head_attention_forward has problem in fp16 mode. in line https://github.com/pytorch/pytorch/blob/5b118a7f233a4682e338c9ef14a5ec55183b9a65/torch/nn/functional.py#L4826 , the product of q and k may have inf in the output and it will lead to nan in the following softmax function. Hope this operation can be more rubost in the fp16 mode.

### PR DESCRIPTION
**Pull request by fuzihaofzh**
Fixes #60939, we detect whether the product is half and contains inf. If it is, we will use float instead of half.

### CODE REVIEW

File: torch/nn/functional.py [hunk 1]
```diff
@@ -4824,12 +4824,15 @@ def _scaled_dot_product_attention(
     q = q / math.sqrt(E)
     # (B, Nt, E) x (B, E, Ns) -> (B, Nt, Ns)
     attn = torch.bmm(q, k.transpose(-2, -1))
+    if attn.isinf().any().item() and attn.dtype is torch.float16:
```
Comment (albanD) [inline torch/nn/functional.py:4827]: This is introducing a sync point as well as recomputing the same thing twice.
I am not convinced this is the best approach.

cc @ngimel

Comment (ngimel) [reply torch/nn/functional.py:4827]: Agree with @albanD, this is a sync point and converting inputs under the hood. Properly scaled models don't result in overflow (there are many examples of transformers training with fp16), so paying this perf penalty always is not an option. cc @ptrblck, @mcarilli.
